### PR TITLE
Re-enable AVX512 ATen kernels for compute-intensive ops

### PR DESCRIPTION
**Pull request by sanchitintel**
## Summary

Enables AVX512 dispatch by default for some kernels, for which AVX512 performs better than AVX2.
For other kernels, their AVX2 counterparts are used.

## Implementation details

`REGISTER_DISPATCH` should now only be used for non-AVX512 dispatch.
`ALSO_REGISTER_AVX512_DISPATCH` should be used when AVX512 dispatch should also be done for a kernel.

## Benchmarking results with #104655 

[Raw data at GitHub Gist (Click on `Download ZIP`)](https://gist.github.com/sanchitintel/87e07f84774fca8f6b767aeeb08bc0c9)

| Op | Speedup of AVX512 over AVX2 |
|----|------------------------------------|
|sigmoid|~27%  with FP32|
|sign| ~16.6%|
|sgn|~15%|
|sqrt|~4%|
|cosh|~37%|
|sinh|~37.5%|
|acos| ~8% with FP32 |
|expm1| ~30% with FP32|
|log|~2%|
|log1p|~16%|
|erfinv|~6% with FP32|
|LogSigmoid|~33% with FP32|
|atan2|~40% with FP32|
|logaddexp| ~24% with FP32|
|logaddexp2| ~21% with FP32|
|hypot| ~24% with FP32|
|igamma|~4% with FP32|
|lgamma| ~40% with FP32|
|igammac|3.5%|
|gelu|~3% with FP32|
|glu|~20% with FP32|
|SiLU|~35% with FP32|
|Softplus|~33% with FP32|
|Mish|~36% with FP32|
|Hardswish|~7% faster with FP32 when tensor can fit in L2 cache|
|Hardshrink|~8% faster with FP32 when tensor can fit in L2 cache|
|Softshrink|~10% faster with FP32 when tensor can fit in L2 cache|
|Hardtanh|~12.5% faster with FP32 when tensor can fit in L2 cache|
Hardsigmoid|~7% faster with FP32 when tensor can fit in L2 cache|
|hypot|~35%|
|atan2|~37%|
|dequantize per channel|~10%|

## Insights gleaned through collected data (future action-items):

1. Inplace variants of some ops are faster with AVX512 although the functional variant may be slower for FP32. Will enable AVX512 dispatch for the inplace variants of such kernels.
2. Almost all BF16 kernels are faster with AVX512, so after PyTorch 2.1 release, will enable AVX512 dispatch for BF16 kernels whose corresponding FP32 kernel doesn't perform well with AVX512.
3. Some kernels rely on auto-vectorization & might perform better with AVX512 once explicit vectorization would be enabled for them.

Data was collected with 26 physical threads of one socket of Intel Xeon 8371HC. Intel OpenMP & tcmalloc were preloaded.

cc @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/cpu/PaddingKernel.cpp
Comment (mingfeima) [inline aten/src/ATen/native/cpu/PaddingKernel.cpp:430]: is it possible to add an default argument to let decide whether this kernel is dispatched to avx512 explicitly ? Probably something like:
```C
// if you want this one to go avx512：
REGISTER_DISPATCH(reflection_pad1d_kernel, &reflection_pad1d_kernel_impl, /* enable_avx512_dispatch */ true);

//  if you want to skip avx512 (use avx2 only), just use the old way：
REGISTER_DISPATCH(reflection_pad1d_kernel, &reflection_pad1d_kernel_impl);
```

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cpu/PaddingKernel.cpp
Comment (jgong5) [reply aten/src/ATen/native/cpu/PaddingKernel.cpp:430]: Agree with @mingfeima and I think it is a better idea to explicitly turn on AVX512 for particular kernels instead of turning off others since there are more kernels to turn off than those to turn on. This can reduces the lines of code changes.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cpu/PaddingKernel.cpp
Comment (sanchitintel) [reply aten/src/ATen/native/cpu/PaddingKernel.cpp:430]: Hi, I had started with such an approach when I first started this task, so that the number of changes in the code could be reduced. It leads to symbol resolution issues.

Comment (jgong5) [reply aten/src/ATen/native/cpu/PaddingKernel.cpp:430]: Is it possible to always register null AVX512 kernels by default while checking if there are already existing registrations (e.g., via some macro def checks)? The real AVX512 kernels can be registered beforehand.

Comment (sanchitintel) [reply aten/src/ATen/native/cpu/PaddingKernel.cpp:430]: Thanks again for your inputs, @jgong5! Yes, such an approach is feasible.

The approach below to improve readability, and decrease the number of changes is also similar in nature to @mingfeima 's suggestion above.

```
// if you want this one to go avx512：
ALSO_REGISTER_AVX512_DISPATCH(reflection_pad1d_kernel, &reflection_pad1d_kernel_impl, /* enable_avx512_dispatch */ true);

//  if you want to skip avx512 (use avx2 only), just use the old way：
REGISTER_DISPATCH(reflection_pad1d_kernel, &reflection_pad1d_kernel_impl);
```

**Comment by sanchitintel** [comment]
Hi @jgong5 & @mingfeima, please check if the new approach looks good.

BTW, @mingfeima, two FP16 UTs are failing. Could this be because of some bugs in their AVX512 implementation? Thanks!

image

**Review by jgong5 — APPROVED** [review]
LGTM now.

**Comment by sanchitintel** [comment]
Thanks again for the review, @jgong5! :)
I'm adding more benchmarks & will mark this PR as `ready for review` today, before seeking feedback from @mingfeima as well. Thanks!

**Comment by sanchitintel** [comment]
Hi @mingfeima, I'll add benchmarking data soon.
Can you please take a look at the new implementation in the meantime? Thanks!

I was wondering if we should rename the macro `REGISTER_DISPATCH` for x86_64 machines to `REGISTER_NON_AVX512_DISPATCH` to prevent any confusion, since I added the macro `ALSO_REGISTER_AVX512_DISPATCH` in this PR.

**Review by mingfeima — APPROVED** [review]
LGTM :)

**Comment by sanchitintel** [comment]
Hi @dzhulgakov,
This PR modifies `aten/src/ATen/native/DispatchStub.h` and `aten/src/ATen/native/DispatchStub.cpp`.
Can you please help review it, since these files [seem to require approval from a core-maintainer/core-reviewer](https://github.com/pytorch/pytorch/blob/main/.github/merge_rules.yaml) for merging? Thank you!

**Review by kit1980 — APPROVED** [review]

**Pull request closed without merging**
